active-rfcs/0000-build-performance.md

```diff
@@ -0,0 +1,187 @@
+- Because of the above, performance has actually regressed in __0.21__, even though it was never the best even before.
+- In order to support SSR in the future we have to move away from page-scanning as the way to find and build assets, since SSR apps by their nature *cannot* be rendered ahead of time.
+
+# Detailed design
+
+## Enforced static use of client directives
+
+The client directives such as `client:load`, `client:idle`, etc will need to be defined in the hydrated component where they are used, and not rendered dynamically. For example the following is __not allowed__:
+
+```astro
+---
+import Clock from '../components/Clock.jsx';
+const { shouldHydrate } = Astro.props;
+const attrs = {
+  'client:idle': shouldHydrate
+};
+---
+<Clock {...attrs} />
+```
+
+We need to know that the site depends on the `client:idle` directive, so that we can *build* the client-side JavaScript needed.
+
+The implementation will be:
+
+1. In the compiler, include the used directive as part of the exported metadata about the component.
+2. In the compiler, mark the component as having statically included a directive.
+  - How to mark is up to the implementer, but we have other metadata attached to hydrated component usage already, and it would make sense to follow this same method.
+3. When rendering, if a component contains a client directive, make sure the directive is matched by the marking in __(2)__.
+4. If the marking is not there, we know that the directive must have been added statically. Throw an `Error` message for this, letting the user know that the directive must be added statically.
+
+## The `local:` directive
+
+A new `local:` directive will be introduced that can be used to set relative links on any element. The syntax looks like this:
+
+```astro
+<img local:src="../images/penguin.png" />
+```
+
+The implementation is:
+
+1. The compiler will see the `local:` directives. They must be used *statically* and not added dynamically (they will be ignored if added dynamically).
+1. The compiler will convert the value of the directive into an import statement. For example the above becomes:
+
+    ```astro
+    ---
+    import imgUrl from "../images/penguin.png";
+    ---
+    <img local:src={imgUrl} />
+    ```
+1. If the value of the directive is *not* a static string, the compiler should ignore that usage and not add an import statement.
+1. The compiler include a warning about the non-static usage of the `local:` directive.
+
+There are a couple of special cases that need to be accounted for:
+
+- `<img srcset>` value is more complex than just a single URL. We need to parse the value and add multiple import statements.
+- A `<link rel=stylesheet>` tag will resolve in an import statement that is side-effectual. As such, we should *remove* the actual link tag in the compiler to prevent it from being rendered.
```

Comment:
say more about this? https://github.com/withastro/rfcs/discussions/28 is making the case to leave `<link rel=stylesheet>` tags completely as-is, without any additional processing. I want to make sure the two RFCs are aligned.

Reply:
This was meant for `local:href`. Like if you do:

```astro
<link rel="stylesheet" local:href="../styles/main.css">
```

It would be equivalent to:

```astro
---
import '../styles/main.css';
---
```

Sound reasonable?

Reply:
ah yea, I like that a lot especially in the context of the style unification RFC!

Maybe rewrite this line to make it more clear that this is what you mean?